lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -679,6 +688,220 @@ void ComponentOp::getAsmBlockArgumentNames(
     setNameFn(block->getArgument(i), ports[i].cast<StringAttr>().getValue());
 }
 
+//===----------------------------------------------------------------------===//
+// CombComponentOp
+//===----------------------------------------------------------------------===//
+
+SmallVector<PortInfo> CombComponentOp::getPortInfo() {
+  auto portTypes = getArgumentTypes();
+  ArrayAttr portNamesAttr = getPortNames(), portAttrs = getPortAttributes();
+  APInt portDirectionsAttr = getPortDirections();
+
+  SmallVector<PortInfo> results;
+  for (size_t i = 0, e = portNamesAttr.size(); i != e; ++i) {
+    results.push_back(PortInfo{portNamesAttr[i].cast<StringAttr>(),
+                               portTypes[i],
+                               direction::get(portDirectionsAttr[i]),
+                               portAttrs[i].cast<DictionaryAttr>()});
+  }
+  return results;
+}
+
+WiresOp calyx::CombComponentOp::getWiresOp() {
+  auto *body = getBodyBlock();
+  auto opIt = body->getOps<WiresOp>().begin();
+  return *opIt;
+}
+
+/// A helper function to return a filtered subset of a comb component's ports.
+template <typename Pred>
+static SmallVector<PortInfo> getFilteredPorts(CombComponentOp op, Pred p) {
+  SmallVector<PortInfo> ports = op.getPortInfo();
+  llvm::erase_if(ports, p);
+  return ports;
+}
+
+SmallVector<PortInfo> CombComponentOp::getInputPortInfo() {
+  return getFilteredPorts(
+      *this, [](const PortInfo &port) { return port.direction == Output; });
+}
+
+SmallVector<PortInfo> CombComponentOp::getOutputPortInfo() {
+  return getFilteredPorts(
+      *this, [](const PortInfo &port) { return port.direction == Input; });
+}
+
+void CombComponentOp::print(OpAsmPrinter &p) {
```

Comment:
Happy to do this for parse, print, and build, but I think verify has so many differences that it would not be worth it to make it a separate function.